### DISCUSSION TITLE
How to prevent a leaf provider to rebuild several times when a root provider is updated?

### DISCUSSION THREAD

**Discussion by gdurandrexel** [rrousselGit/riverpod#3392] [Q&A]
I have a provider A, with 3 providers B1, B2, B3 that watch it, and then a provider C that watches B1, B2 and B3.
If A is updated, C rebuilds 3 times.

Is there a technique to avoid that?
I was thinking perhaps of a buffer provider aggregating the states of B1, B2, B3 and **listening** to them with a debounce duration and updating its state after the duration.
Other thoughts?

**Comment by Colman** [comment]
Ideally, provider C should have no knowledge of provider A. Maybe asses when you want provider C to build in relation to changes in provider B1, B2, B3, then use `select` and `selectAsync` to achieve this. If it's still rebuilding too much, then that means that `B1`, `B2`, or `B3` are rebuilding too much. So you can apply the same princicple and use `select` and `selectAsync` inside of the `B` providers to filter out changes in A.

Keep in mind, that if `selectAsync` is used within provider X, it won't change the loading state of provider X until the provider that it's selecting from has fully re-computed.
